Implement SSE2 SetVector128 HW intrinsics - managed

4creators opened the pull request:
FYI @CarolEidt @fiigii @tannergooding 

~~This is managed implementation only to get early feedback on implementation. Tests will be added a bit later today or actually tomorrow in my time zone.~~

tannergooding commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs:
```diff
@@ -971,47 +972,136 @@ public static class Sse2
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Vector128<sbyte> SetVector128(sbyte e15, sbyte e14, sbyte e13, sbyte e12, sbyte e11, sbyte e10, sbyte e9, sbyte e8, sbyte e7, sbyte e6, sbyte e5, sbyte e4, sbyte e3, sbyte e2, sbyte e1, sbyte e0)
+        {
+            unsafe
```
nit: Put unsafe in the method name

tannergooding replied:
(meaning `public static Vector128<sbyte> SetVector128`)

4creators replied:
Will fix.

tannergooding commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs:
```diff
@@ -971,47 +972,136 @@ public static class Sse2
+        {
+            unsafe
+            {
+                sbyte* buffer = stackalloc sbyte[16];
```
Please log a bug. We should profile this before release.

We should also test out if an unpack sequence (what clang is doing) or if a stack buffer of `int` is better (what gcc is doing)

tannergooding commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs:
```diff
@@ -971,47 +972,136 @@ public static class Sse2
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Vector128<short> SetVector128(short e7, short e6, short e5, short e4, short e3, short e2, short e1, short e0)
+        {
+            Vector128<int> vector1 = ConvertScalarToVector128Int32(e0);
```
Same, we should profile this and determine what the actual "best" sequence is.

4creators replied:
Thnx for review - I have created #16838 to track both performance validation and creation of performance tests for HW intrinsics.

tannergooding commented:
@4creators, are you going to fixup `Sse.SetVector128` as well?

4creators commented:
> are you going to fixup Sse.SetVector128 as well

@tannergooding I have fixed it already https://github.com/dotnet/coreclr/pull/16797/commits/d23d663dc06211ba08e2c3a4df694f1191754e5a and will address all review feedback soon.

4creators commented:
> are you going to fixup Sse.SetVector128 as well

Sorry, missed that it is SSE not SSE2 you are asking for. Sure, will do it while addressing feedback.

4creators commented:
Addressed code review feedback. Due to changes in #16797 this PR should be merged first, than I would need to change #16797 before merging and finally we could merge #16832

@CarolEidt @fiigii @tannergooding PTAL

CarolEidt commented:
Would still like to get a review on this from @tannergooding before merging.

4creators commented:
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

tannergooding commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.cs:
```diff
@@ -467,7 +467,22 @@ public static class Sse
+            // TODO-CQ Optimize algorithm choice based on benchmarks
+
+            // Zero vector and load e2 et index 0
+            Vector128<float> e2Vector = LoadScalarVector128(&e2);
```
@CarolEidt, looking at more existing algorithms, it does look like we will want something that can allow users to use `float`/`double` in these methods (without cost).

It would seem like having an `UnsafeCast` method (or better name) would work well (assuming it also explicitly documents the value of the upper bits as "indeterminate"). -- Maybe we could update `Unsafe.As` to do the right thing, but that would require users to pull in another package to do this stuff...

I had thought about providing additional overloads (such as `UnpackLow(float, float)`), but that still results in "indeterminate" results in some scenarios (such as the two upper bits of the result).

4creators replied:
> Maybe we could update Unsafe.As to do the right thing

Actually this what are you commenting on are some of the most serious pain points in current API surface. We need easy and low cost way to load from memory and move from register any of the supported numerals into lowest position of the vector:

```C#

int value = Foo();
.... 
Vector<int> vector = int.AsVector();
// this may allow to write
Vector<int> vector = UnapckLow(value.AsVector(), value.AsVector());
```

4creators replied:
We could even go further and create helper methods which would expand into corresponding intrinsics like:

```C#

int value = Foo();
Vector<int> vector = value.Shuffle(0);
vector = value.UnpackLow();

// and corresponding vector methods
vector = vector.Shuffle(1);

// what would allow to write
vector = vector.ShiftLeft(4).Shuffle(3).UnpackLow();
```

tannergooding commented on tests/src/JIT/HardwareIntrinsics/X86/Sse2/SetVector128.cs:
```diff
@@ -0,0 +1,303 @@
+
+                    if (Environment.Is64BitProcess)
+                    {
+
```
nit: stray newline

tannergooding commented on tests/src/JIT/HardwareIntrinsics/X86/Sse2/SetVector128.cs:
```diff
@@ -0,0 +1,303 @@
+                        }
+                        catch (PlatformNotSupportedException)
+                        {
+
```
Haven't we been putting comments in the catch block for other similar tests?

tannergooding commented on tests/src/JIT/HardwareIntrinsics/X86/Sse2/SetVector128.cs:
```diff
@@ -0,0 +1,303 @@
+                        byteTable.SetOutArray(result, i);
+                    }
+
+                    CheckMethodSpan<double> checkDouble = (Span<double> x, Span<double> y, Span<double> z, Span<double> a) =>
```
What are `y` and `a` for?

4creators replied:
I have reused some of the test API instead of creating additional methods and for this particular case reused method has too many arguments.

tannergooding commented on tests/src/JIT/HardwareIntrinsics/X86/Sse2/TestTableSse2.cs:
```diff
@@ -224,7 +224,7 @@ public bool CheckResult(CheckMethod<T> check)
         {
             bool result = true;
-            for (int i = 0; i < inArray1.Length; i++)
+            for (int i = 0; i < inArray1.Length; i += ElementCount)
```
Why this change?

4creators replied:
This method was not used before and was created specifically for `Span<T>` arguments. It needs to increment by element count of `Vector<T>`

fiigii commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs:
```diff
@@ -971,47 +972,191 @@ public static class Sse2
+
+            // Zero vector and load e1 et index 0
+            Vector128<double> e1Vector = LoadScalarVector128(&e1);
+            Vector128<double> e0Vector = LoadScalarVector128(&e0);
```
I do not like these memory round-trips (arg-reg -> memory-> vector reg), `SetScalarVector128<T>` should be better.

4creators replied:
Yes, you are right.

4creators commented:
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

4creators commented:
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

4creators commented:
test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx

4creators commented:
One unrelated failure:

```
JIT_Methodical._cctor_misc_threads2_cs_do_threads2_cs_do_._cctor_misc_threads2_cs_do_threads2_cs_do_cmd [FAIL] 
```
in  https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/x64_checked_windows_nt_jitx86hwintrinsicnoavx_prtest/198/

and timout in another job reruning tests

4creators commented:
@tannergooding It is second timeout - should I rerun?

Can we increase timeout to 5h?

```
    Finished:    GC.API.XUnitWrapper
    Finished:    Regressions.common.XUnitWrapper
    Finished:    tracing.tracevalidation.XUnitWrapper
    Finished:    JIT.superpmi.XUnitWrapper
    Finished:    Regressions.coreclr.XUnitWrapper
    Finished:    JIT.SIMD.XUnitWrapper
    Finished:    baseservices.exceptions.XUnitWrapper
    Finished:    JIT.Performance.XUnitWrapper
    Finished:    JIT.CodeGenBringUpTests.XUnitWrapper
    Finished:    JIT.opt.XUnitWrapper
    Finished:    JIT.Generics.XUnitWrapper
    Finished:    JIT.IL_Conformance.XUnitWrapper
    Finished:    JIT.HardwareIntrinsics.XUnitWrapper
    Finished:    GC.Scenarios.XUnitWrapper
    Finished:    JIT.Directed.XUnitWrapper
    Finished:    JIT.jit64.XUnitWrapper
    Finished:    baseservices.threading.XUnitWrapper
    Finished:    JIT.Regression.XUnitWrapper
Build timed out (after 240 minutes). Marking the build as aborted.
Build was aborted
``

tannergooding commented:
> should I rerun

No. The HardwareIntrinsic tests completed without failure and the changes wouldn't impact the other tests.

> Can we increase timeout to 5h?

@jkotas might be able to comment on this...

jkotas commented:
> Can we increase timeout to 5h?

I do not see a problem with increasing the timeout.

@jashook ?

jashook commented:
Sounds good to me, opened PR #16888 to address the issue.
